Colored text for sliders

daynardn opened the pull request:
This is a quick PR I wrote to allow for colored text in sliders different from fg/bg. I'm not sure if this is actually the right approach because it's still calculating the seams for fg and bg just using the same color for both. This also means that a transparent background color works and can still show up when you slide over it.

The documentation says an option for colored text exists yet there's no way to actually use it so this brings parity with the documentation. If this is not a change you want / you have an issue with the code please let me know and I will do my best to fix it.

![image](https://github.com/user-attachments/assets/16a624f0-a79a-4ab8-b2bb-c497e7c659bd)

ogios commented:
wonder if it would be better if we change it to `fg_text_color` and `bg_text_color`, i'm afraid it might introduce complexity but the logic makes more sense.

and thanks, the text color attribute was in use before i implement the splicing of fg&bg colored text, and i just forgot it afterwards...

daynardn commented:
An important note is that for the mic / audio widget, the text doesn't switch colors when you swap into "mute" mode. So if you added `fg_text_color` and `bg_text_color`, you'd probably want to add a `mute_text_color` for that preset? I'm not entirely sure.

ogios commented:
yeah, it's ok to add this, it should be using the same tansition as mute_color to fg color

daynardn commented:
I just added a separate fg and bg color config option. I also added a `mute_text_color` for the pulseaudio widget. You can check this with this widget. If you think that the split fg and bg color is too much please let me know and I can remove it.

```
        {
          "edge": "top",
          "monitor": "eDP-1",
          "position": "left",
          "layer": "overlay",
          "margins": {
            "left": "5%"
          },
          "widget": {
            "type": "slider",
            "thickness": 40,
            "border_width": 6,
            "length": "12.5%",
            "fg_color": "#FF0000",
            "bg_color": "#FF99FFff",
            "obtuse_angle": 100,
            "bg_text_color": "#FFFFaa",
            "fg_text_color": "#FFaaaa",
            "redraw_only_on_internal_update": true,
            "preset": {
                "type": "speaker",
                "mute_text_color": "#000000",
            }
          }
        },
```
Unmuted:
![image](https://github.com/user-attachments/assets/5fa7a7c4-c361-4051-97d5-7fc5f3953269)
Muted:
![image](https://github.com/user-attachments/assets/0895fd46-7be2-45f7-8445-cf1d3fe50b00)

ogios commented:
Thanks, it looks great to me!
i'll update config schema and doc soon after

daynardn commented:
Slight correction to the pr, sorry I was unclear, the second image is wrong. It actually assumes the same color ratio for fg and background rather than the mute color. It completely slipped me my apologies. It's almost the same it's just that if you never set a mute color it picks the original fg / bg ratio.

This is the actual behavior as opposed to the second image which is 
![image](https://github.com/user-attachments/assets/f325de54-d86c-45e2-be70-a3c6cfb06f3a)
 If no text_mute_color is set. 
 
 The same behavior in the image is still possible you just have to set the text_mute_color and it's not default.

ogios commented:
the mute color and mute text color is correct.

`mute color ->(change) fg color`
and so the `mute_text_color ->(change) fg text color`

however, i think what should be changed is `fg_text_color` and `bg_text_color`.
the bg_text_surf should be using the bg_text_color
and for fg the same change.

```rust
// make_text func
        let bg_text_surf = {
            let (bg_text_surf, ctx) = self.new_surface_bar();
            ctx.mask_surface(normal_text_surf, text_start_pos.0, text_start_pos.1)
                .unwrap();

            let (mask_surf, ctx) = self.new_surface_bar();
            // this here
            if let Some(fg_text_color) = conf.fg_text_color {
                cairo_set_color(&ctx, fg_text_color);
            } else {
                cairo_set_color(&ctx, conf.bg_color);
            }

            ctx.mask_surface(&self.fg_surf, Z, Z).unwrap();

            let (final_surf, ctx) = self.new_surface_bar();
            ctx.set_source_surface(mask_surf, Z, Z).unwrap();
            ctx.mask_surface(&bg_text_surf, Z, Z).unwrap();

            final_surf
        };

        let fg_text_surf = {
            let (fg_text_surf, ctx) = self.new_surface_bar();
            // and this here
            if let Some(bg_text_color) = conf.bg_text_color {
                cairo_set_color(&ctx, bg_text_color);
            } else {
                cairo_set_color(&ctx, conf.fg_color);
            }
            
            ctx.mask_surface(normal_text_surf, text_start_pos.0, text_start_pos.1)
                .unwrap();

            // ctx.translate(conf.border_width as f64 + self.text_translate_x, Z);
            // ctx.append_path(&self.bg_path);
            // ctx.set_operator(cairo::Operator::Clear);
            // ctx.fill().unwrap();

            fg_text_surf
        };
```

daynardn commented:
The colors are right but the logic is confusing me currently.

I think the names might be flipped in the code for `bg_text_surf` and `fg_text_surf`. It doesn't make sense you would render the fg color on the fg side. If I understand correctly, fg being the "shaded" section of the bar representing percentage, you want to use the `bg_color` not `fg_color`. I need to look into this more later.

```
// make_text func originally
        let bg_text_surf = {
            let (bg_text_surf, ctx) = self.new_surface_bar();
            ctx.mask_surface(normal_text_surf, text_start_pos.0, text_start_pos.1)
                .unwrap();
                 
                // Why is this bg_color, and not fg_color if we are rendering on the bg? Would this not be the same color?
                cairo_set_color(&ctx, conf.bg_color);
            ...
```

ogios commented:
ok i have that modified, and draft a release
